src/coins.h

```diff
@@ -169,7 +154,16 @@ struct CCoinsCacheEntry
 
     static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel, bool fresh = false) noexcept
     {
-        AddFlags(fresh ? DIRTY | FRESH : DIRTY, pair, sentinel);
+        if (!pair.second.m_flags) {
+            Assume(!pair.second.m_prev && !pair.second.m_next);
+            pair.second.m_prev = sentinel.second.m_prev;
+            pair.second.m_next = &sentinel;
+            sentinel.second.m_prev = &pair;
+            pair.second.m_prev->second.m_next = &pair;
+        }
+        Assume(pair.second.m_prev && pair.second.m_next);
+        pair.second.m_flags |= DIRTY;
+        if (fresh) pair.second.m_flags |= FRESH;
```

Comment:
What should happen when `fresh` is `false`?
Shouldn't we assert something like:
```C++
if (fresh) {
    pair.second.m_flags |= FRESH; 
} else {
    assert(!(pair.second.m_flags & FRESH));
}

Reply:
Should we remove its freshness in that case?

Reply:
But we have prohibited one of the states here, we can document that with code now instead of comments. If we think it's a behavior change the PR is incorrect, isn't it?